src/dist/dist.rs

```diff
@@ -599,6 +600,87 @@ fn update_from_dist_<'a>(
     profile: Option<Profile>,
     prefix: &InstallPrefix,
     force_update: bool,
+) -> Result<Option<String>> {
+    let mut toolchain = toolchain.clone();
+    let mut fetched = String::new();
+    let mut first_err = None;
+    let backtrack = toolchain.channel == "nightly" && toolchain.date.is_none();
+    loop {
+        match try_update_from_dist_(
+            download,
+            update_hash,
+            &toolchain,
+            profile,
+            prefix,
+            force_update,
+            &mut fetched,
+        ) {
+            Ok(v) => break Ok(v),
+            Err(e) => {
+                if !backtrack {
+                    break Err(e);
+                }
+
+                if let ErrorKind::RequestedComponentsUnavailable(components, ..) = e.kind() {
```

Comment:
What if we're now requesting a dated nightly which doesn't exist?  While it's rare, sometimes we don't succeed in building a nightly and so we can't provide one for a given date.

Reply:
How does that manifest in the errors? I can't seem to find an error that maps cleanly to that idea. Basically, how does the manifest change if a nightly is unavailable?

Reply:
@johnthagen current nightly is unavailable so it's perfect occasion to test it :laughing:.

Reply:
Actually, this turns out to already just work, since it manifests as the `rustc` component missing (and thus hits the existing case). I've added a test.

Reply:
Wrong person @mati865